core/src/main/java/io/questdb/cairo/TxnScoreboardV2.java

```diff
@@ -0,0 +1,262 @@
+            if (lockedTxn > UNLOCKED && lockedTxn < txn) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public boolean incrementTxn(int id, long txn) {
+        long internalId = toInternalId(id);
+        assert internalId < entryScanCount;
+        if (Unsafe.getUnsafe().compareAndSwapLong(null, entriesMem + internalId * Long.BYTES, UNLOCKED, txn)) {
+            updateMaxReaderId(internalId);
+            incrementActiveReaderCount();
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean isMax(long txn) {
+        return txn >= getMax();
+    }
+
+    @Override
+    public boolean isRangeAvailable(long fromTxn, long toTxn) {
+        if (getActiveReaderCount() == 0) {
+            return true;
+        }
+
+        long scanCount = readScanCount();
+        for (int i = 0; i < scanCount; i++) {
+            long lockedTxn = Unsafe.getUnsafe().getLongVolatile(null, entriesMem + (long) i * Long.BYTES);
+            if (lockedTxn > UNLOCKED && lockedTxn >= fromTxn && lockedTxn < toTxn) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public boolean isTxnAvailable(long txn) {
+        if (getActiveReaderCount() == 0) {
+            return true;
+        }
+
+        long scanCount = readScanCount();
+        for (int i = 0; i < scanCount; i++) {
+            long lockedTxn = Unsafe.getUnsafe().getLongVolatile(null, entriesMem + (long) i * Long.BYTES);
+            if (lockedTxn == txn) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public long releaseTxn(int id, long txn) {
+        long internalId = toInternalId(id);
+        assert internalId < entryScanCount;
+        Unsafe.getUnsafe().putLongVolatile(null, entriesMem + internalId * Long.BYTES, UNLOCKED);
+        // It's too expensive to count how many readers are left for the txn, caller will have to do additional checks.
+        return 0;
+    }
+
+    public void setTableToken(TableToken tableToken) {
+        this.tableToken = tableToken;
+    }
+
+    private static int toInternalId(int id) {
+        return id + VIRTUAL_ID_COUNT;
+    }
+
+    private long getActiveReaderCount() {
+        return Unsafe.getUnsafe().getLongVolatile(null, activeReaderCountMem);
+    }
+
+    private long getMax() {
+        return Unsafe.getUnsafe().getLongVolatile(null, maxMem);
+    }
+
+    private void incrementActiveReaderCount() {
+        long val;
+        do {
+            val = Unsafe.getUnsafe().getLong(activeReaderCountMem);
+        } while (!Unsafe.getUnsafe().compareAndSwapLong(null, activeReaderCountMem, val, Math.min(0, val) + 1));
+    }
+
+    private long readScanCount() {
+        return Unsafe.getUnsafe().getLongVolatile(null, maxIdMem) + 1;
```

Comment:
Maximum reader id is never decremented, so it's enough for many readers to be opened single time to have as many slots scanned. Considering this, this optimization seems questionable - we should find a way to decrement the maximum reader id.

Reply:
Reader count doesn't go down, true. However, opening 320 readers at once is not easy. Usually, there will be other limits, like connection limits, so it's an extreme case. In all other cases, when the number of readers stays low and re-used, this optimisation works.

Reply:
It's enough to get a single spike in the number of readers to disable this optimization until the next server restart.

To me, @kafka1991 's bitmap approach looks better in this sense. Yes, we do entry_count/8 volatile loads unconditionally, but the subsequent slot reads happen only when the reader is active.

Reply:
I'd propose to leave max reader id as it is, without decrementing it and optimise the implementation in other PRs, like this one #5629 or alternatives.

Reply:
Sounds good. I'd only add a clarifying comment/TODO, so that the code is less confusing.

Reply:
>I t's enough to get a single spike in the number of readers to disable this optimization until the next server restart.

The scoreboards are housekept periodically when there are no active readers, so it has to be a spike and constant use. The price is paid when a new partition version is written and the price in us scale. Given the O3 activity on the table this is not the biggest problem.